iot_relay_server/app/ai_agent.py

```
@@ -65,226 +75,280 @@ def _apply_rules(sensor_data: dict, weather: dict) -> list[dict]:
-        decisions.append(_make_decision("lighting", control_state["lighting"], reason, "low", "llm"))
-
-    if "shading" in parsed:
-        sh = parsed["shading"]
-        control_state["shading"] = {"shade_pct": max(0, min(100, int(sh.get("shade_pct", 0)))), "insulation_pct": max(0, min(100, int(sh.get("insulation_pct", 0))))}
-        decisions.append(_make_decision("shading", control_state["shading"], reason, "low", "llm"))
+    tool_call_traces: list[dict] = []
+
+    for turn in range(MAX_TOOL_TURNS):
+        try:
+            response = await _call_openrouter(messages, tools)
+        except Exception as e:
```

Ruff (0.15.9)

[warning] Do not catch blind exception: `Exception`

(BLE001)